File: Overfitting_Underfitting/Commented/lsm.py
# STEP 0: initialize everything
# note I put '.0' behind every whole number to make sure
# Python interpreter understands that I want floating-point
# precision (the x array are all whole numbers)
x = [2.0, 3.0, 5.0, 6.0, 7.0, 8.0, 9.0, 10.0]  # dataset input
y = [2.0, 4.0, 3.0, 4.5, 6.0, 5.0, 7.6, 6.0]  # dataset output

xAverage = 0  # store average of all input values
yAverage = 0  # store average of all output values

sum1 = 0  # store STEP 3 summation #1
sum2 = 0  # store STEP 3 summation #2

# STEP 1
for i in x:
    xAverage += i  # summation of all values in x
for i in y:
    yAverage += i  # summation of all values in y

# STEP 2
xAverage /= len(x)  # divide by length of x array, finding average
yAverage /= len(y)  # divide by length of y array, finding average

# STEP 3
for i in range(len(x)):
    sum1 += (x[i] - xAverage) * (y[i] - yAverage)
    sum2 += (x[i] - xAverage) * (x[i] - xAverage)

# ...and divide it by sum2
slope = sum1 / sum2  # slope of line

# STEP 4
inter = yAverage - slope * xAverage  # find y-intercept

# print out values
print(sum1)
print(sum2)
print(slope)
print(inter)
print(xAverage)
print(yAverage)

File: Overfitting_Underfitting/Commented/lsm.java
public class lsm {
  public static void main(String[] args) {
    // STEP 0: initialize everything
    // note I put '.0' behind every whole number to make sure the
    // Java interpreter understands that I want floating-point
    // precision.

    // Dataset Input
    double[] x = {
            2.0, 3.0, 5.0, 6.0, 7.0, 8.0, 9.0, 10.0
    };

    // Dataset Output
    double[] y = {
            2.0, 4.0, 3.0, 4.5, 6.0, 5.0, 7.6, 6.0
    };

    // Store average of all input values
    double xAverage = 0;

    // Store average of all output values
    double yAverage = 0;

    // Store STEP 3 summation #1
    double sum1 = 0;

    // Store STEP 3 summation #2
    double sum2 = 0;

    // STEP 1
    for (double val : x) {
      // Summation of all values in x
      xAverage += val;
    }

    for (double val : y) {
      // Summation of all values in y
      yAverage += val;
    }

    // Step 2

    // Divide by length of x array, finding average
    xAverage /= x.length;

    // Divide by length of y array, finding average
    yAverage /= y.length;

    // Step 3
    for (int i = 0; i < x.length; i++) {
      sum1 += (x[i] - xAverage) * (y[i] - yAverage);
      sum2 += (x[i] - xAverage) * (x[i] - xAverage);
    }

    // ...and divide it by sum2
    double slope = sum1 / sum2; // slope of line

    // Step 4
    // Find y-intercept
    double inter = yAverage - slope * xAverage;

    // Print out values
    System.out.println(sum1);
    System.out.println(sum2);
    System.out.println(slope);
    System.out.println(inter);
    System.out.println(xAverage);
    System.out.println(yAverage);
  }
}
